File: leetcode/101-200/132-palindrome-partitioning-ii/bench/min_cut.c
/* Benchmark harness for LeetCode #132 — Palindrome Partitioning II.
 * Mirrors min_cut.kara algorithm-for-algorithm.
 *
 * The palindrome table is a genuinely nested structure (array of row pointers)
 * matching Vec<Vec<bool>> / [][]bool, not a flat n*n block, and its element
 * type is `unsigned char` rather than a packed bitset — a bitset would be a
 * different data structure from the other lanes' bool vectors.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define N 500
#define ITERS 400
#define NP 4

static unsigned char *pal[N];
static long long cut_[N];

static long long min_cut(const char *s) {
    long long n = (long long)strlen(s);
    if (n <= 1) {
        return 0;
    }

    for (long long i = 0; i < n; i++) {
        for (long long j = 0; j < n; j++) {
            pal[i][j] = (i == j);
        }
    }

    for (long long length = 2; length <= n; length++) {
        for (long long lo = 0; lo <= n - length; lo++) {
            long long hi = lo + length - 1;
            int ends_match = (unsigned char)s[lo] == (unsigned char)s[hi];
            int inner_ok = (length == 2) || pal[lo + 1][hi - 1];
            if (ends_match && inner_ok) {
                pal[lo][hi] = 1;
            }
        }
    }

    for (long long i = 0; i < n; i++) {
        cut_[i] = 0;
    }
    for (long long i = 0; i < n; i++) {
        if (pal[0][i]) {
            cut_[i] = 0;
        } else {
            long long best = i;
            for (long long j = 1; j <= i; j++) {
                if (pal[j][i] && (cut_[j - 1] + 1) < best) {
                    best = cut_[j - 1] + 1;
                }
            }
            cut_[i] = best;
        }
    }
    return cut_[n - 1];
}

static void lcg_str(long long seed, long long n, long long alpha, char *out) {
    const char *alphabet = "abcdefghijklmnopqrstuvwxyz";
    long long x = seed;
    long long w = 0;
    for (long long k = 0; k < n; k++) {
        x = (x * 1103515245 + 12345) % 2147483648LL;
        long long target = (x / 65536) % alpha;
        for (long long idx = 0; idx < 26; idx++) {
            if (idx == target) {
                out[w++] = alphabet[idx];
            }
        }
    }
    out[w] = '\0';
}

int main(void) {
    for (long long i = 0; i < N; i++) {
        pal[i] = malloc((size_t)N);
    }

    static char cases[NP][N + 1];
    long long alphas[NP] = {2, 4, 26, 3};
    for (long long j = 0; j < NP; j++) {
        lcg_str(j + 1, N, alphas[j], cases[j]);
    }

    long long sink = 0;
    for (long long it = 0; it < ITERS; it++) {
        long long idx = (it * 3) % NP;
        sink = (sink + min_cut(cases[idx])) % 1000000007LL;
    }
    printf("%lld\n", sink);
    return 0;
}

File: leetcode/101-200/132-palindrome-partitioning-ii/bench/min_cut.rs
// Benchmark harness for LeetCode #132 — Palindrome Partitioning II.
// Mirrors min_cut.kara algorithm-for-algorithm, including the nested
// Vec<Vec<bool>> palindrome table.

fn min_cut(s: &str) -> i64 {
    let bytes = s.as_bytes();
    let n = bytes.len() as i64;
    if n <= 1 {
        return 0;
    }

    let mut pal: Vec<Vec<bool>> = Vec::new();
    for i in 0..n {
        let mut row: Vec<bool> = Vec::new();
        for j in 0..n {
            row.push(i == j);
        }
        pal.push(row);
    }

    let mut length: i64 = 2;
    while length <= n {
        let mut lo: i64 = 0;
        while lo <= n - length {
            let hi = lo + length - 1;
            let ends_match = bytes[lo as usize] == bytes[hi as usize];
            let inner_ok = length == 2 || pal[(lo + 1) as usize][(hi - 1) as usize];
            if ends_match && inner_ok {
                pal[lo as usize][hi as usize] = true;
            }
            lo += 1;
        }
        length += 1;
    }

    let mut cut: Vec<i64> = vec![0; n as usize];
    for i in 0..n {
        if pal[0][i as usize] {
            cut[i as usize] = 0;
        } else {
            let mut best = i;
            let mut j: i64 = 1;
            while j <= i {
                if pal[j as usize][i as usize] && (cut[(j - 1) as usize] + 1) < best {
                    best = cut[(j - 1) as usize] + 1;
                }
                j += 1;
            }
            cut[i as usize] = best;
        }
    }
    cut[(n - 1) as usize]
}

fn lcg_str(seed: i64, n: i64, alpha: i64) -> String {
    let alphabet = "abcdefghijklmnopqrstuvwxyz";
    let mut out = String::new();
    let mut x = seed;
    for _ in 0..n {
        x = (x * 1103515245 + 12345) % 2147483648;
        let target = (x / 65536) % alpha;
        for (idx, ch) in alphabet.chars().enumerate() {
            if idx as i64 == target {
                out.push(ch);
            }
        }
    }
    out
}

fn main() {
    let n: i64 = 500;
    let iters: i64 = 400;

    let cases = [
        lcg_str(1, n, 2),
        lcg_str(2, n, 4),
        lcg_str(3, n, 26),
        lcg_str(4, n, 3),
    ];

    let np = cases.len() as i64;
    let mut sink: i64 = 0;
    for it in 0..iters {
        let idx = ((it * 3) % np) as usize;
        sink = (sink + min_cut(&cases[idx])) % 1000000007;
    }
    println!("{}", sink);
}
